tests/actor/test_run_utils.py

```diff
@@ -0,0 +1,112 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Simple tests for the actor runner."""
+
+import time
+from typing import Iterator
+from unittest.mock import patch
+
+import async_solipsism
+import pytest
+import time_machine
+from async_solipsism.socket import asyncio
+
+from frequenz.sdk.actor import actor, run
+
+
+@pytest.fixture(autouse=True)
+def fake_loop() -> Iterator[async_solipsism.EventLoop]:
+    """Replace the loop with one that doesn't interact with the outside world."""
+    loop = async_solipsism.EventLoop()
+    yield loop
+    loop.close()
+
+
+@pytest.fixture
+def fake_time() -> Iterator[time_machine.Coordinates]:
+    """Replace real time with a time machine that doesn't automatically tick."""
+    with time_machine.travel(0, tick=False) as traveller:
+        yield traveller
+
+
+@actor
+class FaultyActor:
+    """A test faulty actor."""
+
+    def __init__(self, name: str) -> None:
+        """Initialize the test faulty actor.
+
+        Args:
+            name: the name of the test faulty actor.
+        """
+        self.name = name
+        self.is_cancelled = False
+
+    async def run(self) -> None:
+        """Run the test faulty actor.
+
+        Raises:
+            CancelledError: after sleeping the duration in seconds set.
+        """
+        self.is_cancelled = True
+        raise asyncio.CancelledError(f"Faulty Actor {self.name} failed")
+
+
+@actor
+class SleepyActor:
+    """A test actor that sleeps a short time."""
+
+    def __init__(self, name: str, duration_sec: float) -> None:
+        """Initialize the test sleepy actor.
+
+        Args:
+            name: the name of the test sleepy actor.
+            duration_sec: the virtual duration in seconds to run.
+        """
+        self.name = name
+        self.duration_sec = duration_sec
+        self.is_joined = False
+        self.virtual_time_taken_to_run = 0.0
+
+    async def run(self) -> None:
+        """Run the test sleepy actor."""
+        start_time = time.time()
+        await asyncio.sleep(self.duration_sec)
+        self.is_joined = True
+        end_time = time.time()
+        self.virtual_time_taken_to_run = end_time - start_time
+        # pylint: disable=protected-access,no-member
+        await self._stop()  # type: ignore[attr-defined]
+
+
+# pylint: disable=redefined-outer-name
+async def test_all_actors_done(fake_time: time_machine.Coordinates) -> None:
+    """Test the completion of all actors."""
+
+    async def sleep(duration: float) -> None:
+        fake_time.shift(duration)
+
+    with patch("asyncio.sleep", new=sleep):
```

Comment:
Mmm, interesting, but I think this is not the correct approach.

Usually when faking time, you should "drive" the time manually and assert stuff while you make time pass. In this case, since you want to test something (`run`) waiting for something else (the actor run), you probably need a different task to monitor and make the time pass.

So the `await asyncio.sleep(self.duration_sec)` in the actor `run()` won't be executed until you make a `fake_time.shift()` of **at least** `duration_sec`. So you should assert for that to make sure `run()` is really waiting for the actor to finish.

Reply:
Long story short this is my understanding, the `asyncio.sleep` was taken the real time rather than a negligible time given that the test was supposed to use a fake loop. Renaming the function from `fake_loop` to `event_loop` does the trick so that now the time taken is sleep is negligible and does not longer need to be patched.

Moreover we agreed in a call how to better handle the time is god mode :) and verify it. So I still need to amend the tests accordingly.

Also I'll need to patch other tests where `fake_loop` function name was used, will add these changes in a separate commit